lib/xcode/install.rb

```diff
@@ -6,35 +6,108 @@
+    #                        the parameter is a single number that's literally percent (e.g. 1, 50, 80 or 100)
+    # rubocop:disable Metrics/AbcSize
+    def fetch(url: nil,
+              directory: nil,
+              cookies: nil,
+              output: nil,
+              progress: nil,
+              progress_block: nil)
+
       options = cookies.nil? ? [] : ['--cookie', cookies, '--cookie-jar', COOKIES_PATH]
-      # options << ' -vvv'
 
       uri = URI.parse(url)
       output ||= File.basename(uri.path)
       output = (Pathname.new(directory) + Pathname.new(output)) if directory
 
       retry_options = ['--retry', '3']
-      progress = progress ? '--progress-bar' : '--silent'
-      command = ['curl', *options, *retry_options, '--location', '--continue-at', '-', progress, '--output', output, url].map(&:to_s)
+      command = [
+        'curl',
+        *options,
+        *retry_options,
+        '--location',
+        '--continue-at',
+        '-',
+        '--output',
+        output,
+        url
+      ].map(&:to_s)
+
+      progress_log_file = File.join(CACHE_DIR, "progress.#{Time.now.to_i}.progress")
+      FileUtils.rm(progress_log_file) if File.exist?(progress_log_file)
 
       # Run the curl command in a loop, retry when curl exit status is 18
       # "Partial file. Only a part of the file was transferred."
       # https://curl.haxx.se/mail/archive-2008-07/0098.html
       # https://github.com/KrauseFx/xcode-install/issues/210
       3.times do
-        io = IO.popen(command)
-        io.each { |line| puts line }
-        io.close
+        command_string = command.collect(&:shellescape).join(' ')
+
+        # Piping over all of stderr over to a temporary file
+        # the file content looks like this:
+        #  0 4766M    0 6835k    0     0   573k      0  2:21:58  0:00:11  2:21:47  902k
+        # This way we can parse the current %
+        # The header is
+        #  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
+        #
+        # Discussion for this on GH: https://github.com/KrauseFx/xcode-install/issues/276
+        # It was not easily possible to reimplement the same system using built-in methods
+        # especially when it comes to resuming downloads
+        # Piping over stderror to Ruby directly didn't work, due to the lack of flushing
+        # from curl. The only reasonable way to trigger this, is to pipe things directly into a
+        # local file, and parse that, and just poll that. We could get real time updates using
+        # the `tail` command or similar, however the download task is not time sensitive enough
+        # to make this worth the extra complexity, that's why we just poll and
+        # wait for the process to be finished
+
+        command_string += " 2> #{progress_log_file}"
+
+        # Non-blocking call
+        stdin, stdout, stderr, wait_thr = Open3.popen3(command_string)
```

Comment:
use the block form of `popen3`, it will yield these arguments for you

Reply:
Yeah, will do, so I don't have to `.close` all the streams